707 MultipleChoiceQuestion support for attempts

mlippert opened the pull request:
So far I've added support for initializing the mc question w/ the max attempts allowed, which includes updating the BricLayer to recognize the new maxAttempts activity property by copying it into the info domain so it can be referenced by configFixup.

I've also added support for getting and updating the attempts made from the attemptsMade property in the response object.

The submit button is correctly left disabled when the submitted answer was correct or if it was the last allowable attempt.

Currently there is untested code to display the correct answer when the last allowed attempt was incorrect. This needs the LocalAnswerMan to be updated to handle attempts, in order to test, which is the next thing I intend to do.

Commit:
tabs to spaces

Commit:
support the activityConfig optional maxAttempts property

Commit:
Initial work towards supporting maxAttempts on the MultipleChoiceQuestion bric

Commit:
added attempts handling to multiplechoicequestion

because this required depending on more functionality from the presenter bric, I created a new interface IChoicePresenter.

That required that I implement that class in RadioGroup and (to some extent) in SelectGroup.

Commit:
Fix eventmanager to publish to all subscribers when some are unsubscribed during event

Commit:
update existing unit tests for multiplechoicequestion to pass

Commit:
tabs to spaces

ysahnpark commented:
Cool.
Once done, I'd be happy to try it out in the integration demo.
I guess changes in LocalAnswerMan will not cause conflicts with changes I did on IpsAnswerMan.

Commit:
Merge branch 'origin/master'

Commit:
split the localanswerman into its own file

Commit:
add attemptsMade and maxAttempts

actually using maxAttempts to determine when to add the correctAnswer will come later.

Commit:
add an alwayscorrect evaluator to the local answerman

Commit:
add pearson.utils.logToConsole function

Commit:
local answerman now supports max attempts and returning correct answer

Commit:
cleanup the score request objects from the question to the scoring engine and back

This meant renaming various properties.

In the score request event details:
* question was removed
* questionId -> submissionId
* answerKey was a string specific to multiple choice. it was replaced by the answer object, which for multiple choice has a key property containing the old answerKey value.

In the response details:
* score -> correctness
* response -> feedback

mlippert commented:
Here's a synopsis of what's been done. I think the only thing left is unit tests.

## answerman.js
- extracted `LocalAnswerMan` to `localanswerman.js`
- in `IpsAnswerMan` changed callback arg property names
  - score -> correctness
  - response -> feedback

## briclayer.js
- handle new activity config property `maxAttempts`

## briclayer-spec.js
- add unit test of new config property `maxAttempts`

## eventmanager.js
- change to allow the executing event handler to unsubscribe itself

## localanswerman.js
- created from `answerman.js`
- added support for `maxAttempts`
  - settable property - setMaxAttempts(maxA)
  - track attemptsMade
  - return attemptsMade
  - created functions to get the correctAnswer
  - return correctAnswer when '`attemptsMade === maxAttempts`' and the answer was incorrect
  - added support for question type `alwayscorrect`
  - changed response property names
    - score -> correctness
    - response -> feedback

## submitmanager.js

### handleScoreRequest_/handleScoringResponse_

The event details structure has changed
- questionId -> submissionId
- `answerKey` was a string specific for multiple choice submissions. It is gone, replaced by an `answer` property which is an object. For multiple choice the old `answerKey` string is now found at `answer.key`.

### appendResponseWithDefaultFormatting

The response details structure has changed
- score -> correctness
- response -> feedback
- The code processing correctness expects a float value from 0 - 1, instead of an integer -1 - 2

## Template_Multiple_Choice_Activity.html
- Added maxAttempts property
- set Q1's config maxAttempts property
- set LocalAnswerMan's maxAttempts property

## widget-base.js
- add `pearson.utils.logToConsole` function
- add `pearson.brix.KeyedAnswer` type (for multiple choice)
- add `pearson.brix.IChoicePresenter` interface (perhaps it should be a base class derived from HtmlBric?)

## widget-multiplechoicequestion.js
- add logger_ {goog.debug.Logger}
- change `window.console.log` to use `logger_`
- add `maxAttempts` to the config default of unlimited
- add maxAttempts and attemptsMade handling
- change initial button text to 'Submit'
- unsubscribe from answer selected event after the initial selection
- changed event details structure published w/ the submitScoreRequest event
  - removed question
  - questionId -> submissionId
  - answerKey -> answer.key
- disabled the submit button when submitting the score request (the response processing may re-enable it)
- changed response processing to handle new response structure, and logic for maxAttempts/attemptsMade
- add rendering of attemptsMade/maxAttempts

## widget-radiogroup.js
- implement `IChoicePresenter` interface
  - except `flagChoice` method was left doing nothing. It should eventually overwrite the radio button next to the given choice w/ an green check icon.

## widget-selectgroup.js
- implement `IChoicePresenter` interface so the `ManzaHealth.html` demo continues to work.

## widgets.css
- add preliminary formatting for multiple choice question w/ radiogroup presenter

Commit:
fix logic for adding correct answer to test correctness property not score

Commit:
correctAnswer property in score request response may exist and be null

Commit:
fix maxAttempts === null means unlimited attempts

Commit:
add multiple choice and radiogroup unit test placeholders

Commit:
add localanswerman.js to multiplechoice integration test html file

Commit:
submission response correctAnswer property should not have a correctness property because it will always have a value of 1

The server was doing this, but I had forgotten when I implemented the localanswerman and multiple choice bric, so it showed up when I ran the integration test.

mlippert commented:
:+1: Although the unit tests have not been implemented for this code, @lbondaryk, @ysahnpark and @seannives have approved it being merged to master, for the sake of expediency.

I have created some placeholders for tests that should be implemented on radiogroup and multiplechoicequestion, and all existing unit tests have been updated to pass.

I've also run the multiplechoice integration test and fixed the issues that it revealed except the server issue (it is returning the correctAnswer on the last attempt when the last attempt was correct).
